checker/src/org/checkerframework/checker/index/qual/IndexOfIndexFor.java

```diff
@@ -0,0 +1,52 @@
+package org.checkerframework.checker.index.qual;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import org.checkerframework.framework.qual.JavaExpression;
+import org.checkerframework.framework.qual.SubtypeOf;
+
+/**
+ * The annotated expression evaluates to either -1 or a non-negative integer less than the lengths
+ * of all the given sequences.
+ *
+ * <p>This is the return type of {@link java.lang.String#indexOf(String) String.indexOf} and {@link
+ * java.lang.String#lastIndexOf(String) String.lastIndexOf} in the JDK. The return type of these
+ * methods is annotated {@code @IndexOfIndexFor(value="this",offset="#1.length()-1")}, meaning that
+ * the returned value is either -1 or it is less than or equal to the length of the receiver
+ * sequence minus the length of the sequence passed as the first argument.
+ *
+ * <p>The annotation {@code @IndexOfIndexFor(args)} is like <code>
+ * {@literal @}{@link NonNegative} {@literal @}{@link LTLengthOf}(args)</code>, except that
+ * {@code @IndexOfIndexFor(args)} additionally permits the value -1.
+ *
+ * <p>The name of this annotation, "indexOf index for", is intended to mean that the annotated
```

Comment:
I don't feel that this conveys the intended meaning.  The annotated expression is not necessarily an index for the given sequence, because it might be -1.  The documentation calls it an "indexOf result", which would suggest an annotation named `@IndexOfResult`.  Can you think of a better name than that?

Reply:
Another possibility would be something like `@IndexOfSubstring`.

Reply:
Do you have any thoughts on naming?

Reply:
I might call it `SubstringIndex` to extend the naming convention that we used for `SearchIndex`. In general, I'm partial to a name that includes `substring` somehow, because I think that most effectively conveys its intended use.

Reply:
`SubstringIndex` is fine with me.  Thanks for the idea.

Reply:
Thank you, I think it is a good suggestion, so the hierarchy would be `SubstringIndexUnknown`, `SubstringIndexFor(value, offset)` (or alternatively `IndexOfSubstring(value,in)`) and `SubstringIndexBottom`.
Should I rename the whole checker to org.checkerframework.checker.index.substringindex.SubstringIndexChecker?

Reply:
I think the names should be consistent, so I wouldn't use both `SubstringIndexUnknown` and `IndexOfSubstring(value,in))`, and I would rename the checker for consistency as well.